File: DSA/Pranvkumar_590011587/Day23_Question3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

# Delete Middle of Linked List
def delete_middle_node(head):
    if not head or not head.next:
        return None

    slow = head
    fast = head
    prev = None

    while fast and fast.next:
        fast = fast.next.next
        prev = slow
        slow = slow.next

    if prev:
        prev.next = slow.next
    return head

def print_list(head):
    current = head
    while current:
        print(current.data, end=" -> " if current.next else "\n")
        current = current.next

# Sample Usage:
list = LinkedList()
for val in [1, 2, 3, 4, 5]:
    list.append(val)
print("Original list:")
print_list(list.head)
head3 = delete_middle_node(list.head)
print("After deleting middle:")
print_list(head3)

File: DSA/upama_kumari_590014909/day_23_question_3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node* deleteMiddle(struct Node* head) {
    if (head == NULL || head->next == NULL) {
        free(head);
        return NULL;
    }

    struct Node *slow = head, *fast = head, *prev = NULL;

    while (fast != NULL && fast->next != NULL) {
        fast = fast->next->next;
        prev = slow;
        slow = slow->next;
    }

    prev->next = slow->next;
    free(slow);

    return head;
}

void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d", head->data);
        if (head->next) printf(" -> ");
        head = head->next;
    }
    printf("\n");
}

struct Node* createExampleList() {
    struct Node* head = newNode(2);
    head->next = newNode(4);
    head->next->next = newNode(6);
    head->next->next->next = newNode(7);
    head->next->next->next->next = newNode(5);
    head->next->next->next->next->next = newNode(1);
    return head;
}

int main() {
    struct Node* head = createExampleList();

    printf("Original List:\n");
    printList(head);

    head = deleteMiddle(head);

    printf("After Deleting Middle:\n");
    printList(head);

    return 0;
}
